pallets/collator-selection/src/lib.rs

```diff
@@ -330,15 +330,40 @@ pub mod pallet {
 		pub fn account_id() -> T::AccountId {
 			T::PotId::get().into_account()
 		}
+		/// Removes a candidate if they exist and sends them back their deposit
+		fn try_remove_candidate(who: &T::AccountId) -> Result<usize, DispatchError> {
+			let current_count = <Candidates<T>>::try_mutate(|candidates| -> Result<usize, DispatchError> {
+				let index = candidates.iter().position(|candidate| candidate.who == *who).ok_or(Error::<T>::NotCandidate)?;
+				T::Currency::unreserve(&who, candidates[index].deposit);
+				candidates.remove(index);
+				Ok(candidates.len())
+			});
+			Self::deposit_event(Event::CandidateRemoved(who.clone()));
+			current_count
+		}
 
 		/// Assemble the current set of candidates and invulnerables into the next collator set.
 		///
 		/// This is done on the fly, as frequent as we are told to do so, as the session manager.
 		pub fn assemble_collators() -> Vec<T::AccountId> {
 			let mut collators = Self::invulnerables();
+			let boot_block = Self::boot_block();
 			collators.extend(
-				Self::candidates().into_iter().map(|c| c.who).collect::<Vec<_>>(),
+				Self::candidates().into_iter().filter_map(|c| {
+					if c.last_block > boot_block {
```

Comment:
I probably recommended this myself, but this is wrong. What you need is:

```
let now = system::now();
let boot_threshold = T::boot();
let last = c.last_block;
let since_last = now - last;

if since_last > boot { ... }
```

Reply:
So no I don't think this is, What is happening is lower in the function KickBlock (oh ya boot block, I just renamed it will push soon) gets set (it would start as 0). Then a session happens block is now X, if a collator produced a block in the time they will be greater then kickBlock else they will be kicked then kick block gets moved up again

Reply:
I see, then it is semantically the same as why I recommended, right? except, in your way we have to read and write into a storage item, but you can do the same if you make `KickBlock` a config `Get`. Or am I missing something?

Reply:
ya, but to avoid config confustion Id want to make the kick_block the session block, this is becuase this only runs on a new session, which makes sense to me because that is when we set new authoritites.